### PR TITLE
Cache `myReduced = NoType` if match type reduction overflowed

### PR DESCRIPTION
**Pull request by EugeneFlesselle**

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 1]
```diff
@@ -5124,6 +5124,7 @@ object Types extends TypeUtils {
       }
 
       def isUpToDate: Boolean =
+        (reductionContext ne null) &&
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/Types.scala:5127]: != null is fine, it's the same code generated.

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 2]
```diff
@@ -5169,21 +5170,20 @@ object Types extends TypeUtils {
       then
         record("MatchType.reduce computed")
         if (myReduced != null) record("MatchType.reduce cache miss")
-        myReduced =
-          trace(i"reduce match type $this $hashCode", matchTypes, show = true):
+        if !isUpToDate then setReductionContext()
+        val saved = ctx.typerState.snapshot()
+        try
+          myReduced = trace(i"reduce match type $this $hashCode", matchTypes, show = true):
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/Types.scala:5176]: I see now: We did cache reduce failures before _except those that threw an exception_. That's handled now.

File: compiler/src/dotty/tools/dotc/core/Types.scala (hunk 1 above)
Comment (EugeneFlesselle) [reply compiler/src/dotty/tools/dotc/core/Types.scala:5127]: Except values of type `MutableMap[Type, Type]` and `Null` cannot be compared with `!=`. 

Is there an advantage to doing 
```scala
private var reductionContext: util.MutableMap[Type, Type] = uninitialized
def setReductionContext(): Unit =
  ...
  reductionContext = util.HashMap()
 ```
rather than 
```scala
private var reductionContext: util.MutableMap[Type, Type] = util.HashMap()
def setReductionContext(): Unit =
  ...
  reductionContext.clear()
```
?

Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/Types.scala:5127]: I overlooked that. The proper thing to do is to declare reductionContext as 

```scala
    private var reductionContext: util.MutableMap[Type, Type] | Null = null
```
Generally, `= unintialized` should only be used if we never need to test for null afterwards since the logic ensures that the var is always assigned before use.

**Review by odersky — COMMENTED** [review]
Last changes look good to me
